### PR TITLE
Node/Acct: audit stuck pending

### LINKED ISSUE #2002: Re-observations for transfers stuck in global accountant
Since accounting happens before a VAA is broadcasted over the p2p network we cannot use the existing re-observation infrastructure to unblock transfers that get stuck in accounting waiting for quorum.

However, since we know exactly which signatures are missing we also don't need to spam the entire p2p network for re-observations.  Instead we can have each guardian query the accounting contract for any pending transfers for which it doesn't have an observation from that guardian.  The guardian can then re-observe only those txs and send the observations once it has them.

The proposal is to add a new query method to the accounting contract:

```rust

#[cw_serde]
pub enum QueryMsg {
    /// Query the contract for any pending observations for which it doesn't have an observation
    /// from the provided guardian index.
    #[returns(QueryMissingObservationsResponse)]
    MissingObservations {
        /// The index of the guardian whose signatures we want to check.
        index: u8,
    },
}

#[cw_serde]
pub struct QueryMissingObservationsResponse {
    /// A list of transfers for which the accounting contract does not have an observation from the
    /// requested guardian, along with a list of possible tx_hashes where that transfer may have
    /// occurred.
   missing_observations: Vec<transfer::Key, Vec<Binary>>,
}
```

guardiand would then send this query every (5? 7? 11?) minutes and try to re-observe any missing observations.

### PR DESCRIPTION
**Pull request by bruce-riley**
This PR implements the global account audit in the guardian. It verifies that the set of pending transfers the guardian knows about matches what the smart contract thinks. If mismatches are detected, it will either resubmit the observation to the contract or request a reobservation from the local watcher.

Additionally, this PR cleans up a couple of minor issues left over from the previous code review.

Fixes #2002 

Change-Id: I643ce699b5d5e21129d957dae8677e8f2fccdbfd

### CODE REVIEW

File: node/pkg/accountant/audit.go [hunk 1]
```diff
@@ -0,0 +1,292 @@
...
+	acct.pendingTransfersLock.Lock()
+	defer acct.pendingTransfersLock.Unlock()
+
+	chanId := vaa.ChainID(mo.ChainId)
+	txHash := ethCommon.BytesToHash(mo.TxHash)
+	for _, pe := range acct.pendingTransfers {
+		if (pe.msg.EmitterChain == chanId) && (pe.msg.TxHash == txHash) {
+			return true
+		}
+	}
```
Comment (jynnantonix) [inline node/pkg/accountant/audit.go:233]: I'm not sure about this.  We know the transfer wasn't in the pending map when we started the audit and it feels unlikely to have been added in the short time since the audit was started so we'll end up iterating over the entire map for every missing observation.

The other end of this channel already deduplicates reobservation requests so maybe we don't need to do this check here.  See https://github.com/wormhole-foundation/wormhole/blob/fac7e10eb4ba195adcfae74b78e7771229650adb/node/cmd/guardiand/reobserve.go#L22-L25

File: node/pkg/accountant/audit.go [hunk 2]
```diff
@@ -0,0 +1,292 @@
...
+
+	var response BatchTransferStatusResponse
+	if err := json.Unmarshal(resp.Data, &response); err != nil {
+		return nil, fmt.Errorf("failed to unmarshal response: %w", err)
+	}
+
+	ret := make(map[string]TransferDetails)
+	for _, item := range response.Details {
+		ret[item.Key.String()] = item
```
Comment (jynnantonix) [inline node/pkg/accountant/audit.go:287]: nit: `ret[item.Key.String()] = item.Status` ?

File: node/pkg/accountant/query_test.go [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func TestParseMissingObservationsResponse(t *testing.T) {
+	//TODO: Write this test once we get a sample response.
+}
+
+func TestParseBatchTransferStatusResponse(t *testing.T) {
```
Comment (jynnantonix) [inline node/pkg/accountant/query_test.go:21]: I think we should also include a test for the case where the contract doesn't know about a transfer and returns a null status.

File: node/pkg/accountant/submit_obs.go [hunk 4]
```diff
@@ -86,7 +88,26 @@ func readFromChannel[T any](ctx context.Context, ch <-chan T, count int) ([]T, e
 	return out, nil
 }
 
+// removeCompleted drops any messages that are no longer in the pending transfer map. This is to handle the case where the contract reports
+// that a transfer is committed while it is in the channel. There is no point in submitting the observation once the transfer is committed.
+func (acct *Accountant) removeCompleted(msgs []*common.MessagePublication) []*common.MessagePublication {
```
Comment (jynnantonix) [inline node/pkg/accountant/submit_obs.go:93]: `acct.pendingTransfersLock.Lock()` ?

File: node/pkg/accountant/audit.go [hunk 5]
```diff
@@ -0,0 +1,292 @@
...
+func (acct *Accountant) handleMissingObservation(mo MissingObservation) {
+	// It's possible we received this transfer after we built the temporary map. If so, we don't want to do a reobservation.
+	if acct.transferNowExists(mo) {
+		acct.logger.Debug("acctaudit: contract reported unknown observation as missing but it is now in our pending map, ignoring it", zap.Uint16("chainId", mo.ChainId), zap.String("txHash", hex.EncodeToString(mo.TxHash)))
+		return
+	}
+
+	acct.logger.Debug("acctaudit: contract reported unknown observation as missing, requesting reobservation", zap.Uint16("chainId", mo.ChainId), zap.String("txHash", hex.EncodeToString(mo.TxHash)))
+	msg := &gossipv1.ObservationRequest{ChainId: uint32(mo.ChainId), TxHash: mo.TxHash}
+	acct.obsvReqWriteC <- msg
```
Comment (hendrikhofstadt) [inline node/pkg/accountant/audit.go:219]: potentially blocking

File: node/pkg/accountant/accountant.go [hunk 6]
```diff
@@ -306,7 +273,7 @@ func (acct *Accountant) deletePendingTransfer(msgId string) {
 func (acct *Accountant) deletePendingTransferAlreadyLocked(msgId string) {
 	acct.logger.Debug("acct: deletePendingTransfer", zap.String("msgId", msgId))
 	if _, exists := acct.pendingTransfers[msgId]; exists {
-		transfersOutstanding.Dec()
+		transfersOutstanding.Set(float64(len(acct.pendingTransfers)))
 		delete(acct.pendingTransfers, msgId)
```
[the remark this answers is not in the document]
Comment (bruce-riley) [reply node/pkg/accountant/accountant.go:277]: Good catch. Fixed.

File: node/pkg/wormconn/query.go [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+
+import (
+	"context"
+	"fmt"
+
+	wasmdtypes "github.com/CosmWasm/wasmd/x/wasm/types"
+)
+
+// SubmitQuery submits a query to a smart contract and returns the result.
+func (c *ClientConn) SubmitQuery(ctx context.Context, contractAddress string, query []byte) (*wasmdtypes.QuerySmartContractStateResponse, error) {
```
[the remark this answers is not in the document]
Comment (bruce-riley) [reply node/pkg/wormconn/query.go:11]: Changed.

File: node/pkg/accountant/audit.go [hunk 8]
```diff
@@ -0,0 +1,292 @@
...
+	missingObservations, err := acct.queryMissingObservations()
+	if err != nil {
+		acct.logger.Error("acctaudit: unable to perform audit, failed to query missing observations", zap.Error(err))
+		for _, pe := range tmpMap {
+			acct.logger.Error("acctaudit: unsure of status of pending transfer due to query error", zap.String("msgId", pe.msgId))
+		}
+		return
+	}
+
+	if len(missingObservations) != 0 {
```
[the remark this answers is not in the document]
Comment (bruce-riley) [reply node/pkg/accountant/audit.go:135]: Old C/C++ habit where you didn't want to set up a for loop if it wasn't going to execute.

I removed the check.

File: node/pkg/accountant/submit_obs.go (hunk 4 above)
Comment (bruce-riley) [reply node/pkg/accountant/submit_obs.go:93]: Fixed.

File: node/pkg/accountant/query_test.go (hunk 3 above)
Comment (bruce-riley) [reply node/pkg/accountant/query_test.go:21]: It was there, but I split it into separate tests to make it more obvious.

File: node/pkg/accountant/audit.go [hunk 9]
```diff
@@ -0,0 +1,292 @@
...
+	defer acct.pendingTransfersLock.Unlock()
+
+	if time.Since(acct.lastAuditTime) < auditInterval {
+		acct.logger.Debug("acctaudit: in AuditPendingTransfers, not time to run yet", zap.Stringer("lastAuditTime", acct.lastAuditTime))
+		return
+	}
+
+	tmpMap := make(map[string]*pendingEntry)
+	for _, pe := range acct.pendingTransfers {
+		if (pe.submitPending) && (time.Since(pe.updTime) > maxSubmitPendingTime) {
```
[the remark this answers is not in the document]
Comment (bruce-riley) [reply node/pkg/accountant/audit.go:108]: Removed them.

File: node/pkg/accountant/audit.go [hunk 10]
```diff
@@ -0,0 +1,292 @@
...
+			for _, pe := range tmpMap {
+				acct.logger.Error("acctaudit: unsure of status of pending transfer due to query error", zap.String("msgId", pe.msgId))
+			}
+			return
+		}
+
+		for _, pe := range pendingTransfers {
+			item, exists := transferDetails[pe.msgId]
+			if !exists {
+				if !pe.submitPending {
```
[the remark this answers is not in the document]
Comment (bruce-riley) [reply node/pkg/accountant/audit.go:175]: I was doing the check separately to avoid unnecessary error logging. I reworked it so that submitObservation indicates if it did the submit so I can control logging.

File: node/pkg/accountant/audit.go (hunk 1 above)
Comment (bruce-riley) [reply node/pkg/accountant/audit.go:233]: Yeah, I guess it is a pretty small window. I removed the check.

File: node/pkg/accountant/audit.go (hunk 2 above)
Comment (bruce-riley) [reply node/pkg/accountant/audit.go:287]: Changed.

File: node/pkg/accountant/audit.go (hunk 5 above)
Comment (bruce-riley) [reply node/pkg/accountant/audit.go:219]: Fixed.

File: node/pkg/accountant/audit.go [hunk 11]
```diff
@@ -0,0 +1,292 @@
...
+				} else {
+					digestMismatches.Inc()
+					acct.logger.Error("acctaudit: audit detected a digest mismatch, dropping transfer", zap.String("msgId", pe.msgId), zap.String("ourDigest", pe.digest), zap.String("reportedDigest", digest))
+					acct.deletePendingTransfer(pe.msgId)
+				}
+			} else if item.Status.Pending != nil {
+				acct.logger.Debug("acctaudit: contract says transfer is still pending", zap.String("msgId", pe.msgId))
+			} else if !pe.submitPending {
+				auditErrors.Inc()
+				acct.logger.Error("acctaudit: contract does not know about pending transfer, resubmitting it", zap.String("msgId", pe.msgId))
```
[the remark this answers is not in the document]
Comment (bruce-riley) [reply node/pkg/accountant/audit.go:200]: This is the case where the contract does not know about a transfer. I added a comment to document that.

It shouldn't happen, which is why I'm logging an error.

File: node/pkg/accountant/accountant.go [hunk 12]
```diff
@@ -344,18 +311,25 @@ func (acct *Accountant) loadPendingTransfers() error {
 // submitObservation sends an observation request to the worker so it can be submited to the contract.
 // If writing to the channel would block, this function resets the timestamp on the entry so it will be
 // retried next audit interval. This method assumes the caller holds the lock.
-func (acct *Accountant) submitObservation(msg *common.MessagePublication) {
+func (acct *Accountant) submitObservation(pe *pendingEntry) {
+	pe.submitPending = true
```
[the remark this answers is not in the document]
Comment (bruce-riley) [reply node/pkg/accountant/accountant.go:315]: Created a state struct in the pendingTransfer and added a lock for it.

File: node/pkg/accountant/audit.go [hunk 13]
```diff
@@ -0,0 +1,292 @@
...
+			auditErrors.Inc()
+			acct.logger.Error("acctaudit: transfer has been in the submit pending state for too long", zap.Stringer("lastUpdateTime", pe.updTime))
+		}
+		acct.logger.Debug("acctaudit: will audit pending transfer", zap.String("msgId", pe.msgId), zap.Stringer("lastUpdateTime", pe.updTime))
+		tmpMap[pe.makeAuditKey()] = pe
+	}
+
+	acct.logger.Debug("acctaudit: in AuditPendingTransfers: starting audit", zap.Int("numPending", len(tmpMap)))
+	acct.lastAuditTime = time.Now()
+	go acct.performAudit(tmpMap)
```
[the remark this answers is not in the document]
Comment (bruce-riley) [reply node/pkg/accountant/audit.go:118]: Changed it to use a runnable with a ticker.

File: node/pkg/accountant/submit_obs.go [hunk 14]
```diff
@@ -173,6 +188,7 @@ func (sb SignatureBytes) MarshalJSON() ([]byte, error) {
 // It should be called from a go routine because it can block.
 func (acct *Accountant) submitObservationsToContract(msgs []*common.MessagePublication, gsIndex uint32, guardianIndex uint32) {
 	txResp, err := SubmitObservationsToContract(acct.ctx, acct.logger, acct.gk, gsIndex, guardianIndex, acct.wormchainConn, acct.contract, msgs)
+	acct.clearSubmitPendingFlags(msgs)
```
[the remark this answers is not in the document]
Comment (bruce-riley) [reply node/pkg/accountant/submit_obs.go:191]: I moved this call to the end of the function (and all the error returns).

File: node/pkg/accountant/accountant.go [hunk 15]
```diff
@@ -283,15 +262,15 @@ func (acct *Accountant) publishTransferAlreadyLocked(pe *pendingEntry) {
 }
 
 // addPendingTransferAlreadyLocked adds a pending transfer to both the map and the database. It assumes the caller holds the lock.
-func (acct *Accountant) addPendingTransferAlreadyLocked(msgId string, msg *common.MessagePublication, digest string) error {
-	acct.logger.Debug("acct: addPendingTransferAlreadyLocked", zap.String("msgId", msgId))
-	if err := acct.db.AcctStorePendingTransfer(msg); err != nil {
+func (acct *Accountant) addPendingTransferAlreadyLocked(pe *pendingEntry) error {
+	acct.logger.Debug("acct: addPendingTransferAlreadyLocked", zap.String("msgId", pe.msgId))
+	pe.state.updTime = time.Now()
```
Comment (jynnantonix) [inline node/pkg/accountant/accountant.go:267]: I assume we should also acquire `pe.stateLock` here?  The caller only holds `acct.pendingTransfersLock` which doesn't protect against modification to the `pendingEntry` state.

**Review by jynnantonix — DISMISSED** [review]

**Review by hendrikhofstadt — APPROVED** [review]

File: node/pkg/accountant/accountant.go (hunk 15 above)
Comment (bruce-riley) [reply node/pkg/accountant/accountant.go]: Fixed.

**Review by hendrikhofstadt — APPROVED** [review]

**Review by hendrikhofstadt — APPROVED** [review]

**Review by jynnantonix — APPROVED** [review]
